### PR TITLE
Update the size of the pre-industrial ocean carbon pools. 

### PR DESCRIPTION
**Pull request by kdorheim**
This is a WIP but needs help from @bpbond! 

## Objective 

The objective of this PR is to update the pre-industrial ocean carbon pools, previously they were undocumented. Our approach was to use values from IPCC AR6 Figure 5.12 to set up the carbon pools and split proportionally with respect to the relative volumes of the Hector ocean boxes. 

## What has changed 

image

Older Hector values 
* HL = 140
* LL = 770
* inter = 8400
* deep = 26000
* total = 35310

New Hector values 
* HL = 135
* LL = 765 
* inter = 9080
* deep = 28019
* total = 37999

The change in pre-industrial C values does cause changes in Hector output values 

image

image

image

Because each of these variables has a different scale I wanted to take a look at the percent change we observe in each of the variables/scenarios when the carbon parameter values are changed. 

image

The mean percent change 

|variable       |       mean|
|:--------------|----------:|
|atm_ocean_flux |  5.5876625|
|DIC_HL         |  7.5413844|
|DIC_LL         |  7.6440605|
|ocean_c        |  7.5970685|
|pH_HL          |  0.4603870|
|pH_LL          |  0.4296637|
|Tgav           | -1.3648476|

The largest percent change we see is in the ocean carbon content  ~ 7%, ph which is log scaled has a small percent change with a 1% change in temperature. 

The question of are these Hector output variables consistent with ESMs/historical observations is a good question and remains to be addressed. There is a problem to be addressed beforehand. Please see the next post for a description of the problems encountered while working on this.

### CODE REVIEW

**Comment by cahartin** [comment]
@kdorheim re documentation, the BGC paper has a description of this in more detail.  But essentially, the ocean carbon cycle was calculated such that there are approximately 100 Pg of carbon are flowing from the high latitude box to the deep box.  Initial values for each box were constructed similar to [Lenton 2000](https://www.tandfonline.com/doi/pdf/10.3402/tellusb.v52i5.17097?needAccess=true).  Then as the carbon cycle gets spun up those values shift a bit to come into equilibrium with the system.  With the new values in place is the carbon flow still around 100 Pg? I suspect so since the volume transports weren't changed.

**Comment by kdorheim** [comment]
# Probelms 

* Carbon parameters are hardcoded in the previous version, although values can technically be read in via the ini file any user-provided parameter is overwritten by the hardcoded values see issue #530 for more details, note that this is a v3.5 bug to be fixed. 
* Unknown variable checks - in the ocean_cycs check there are two assert statements that for [dic and alk](https://github.com/JGCRI/hector/blob/3adda950c9181c95072c763443c0f347d2b3a91f/src/ocean_csys.cpp#L169https://github.com/JGCRI/hector/blob/3adda950c9181c95072c763443c0f347d2b3a91f/src/ocean_csys.cpp#L169) that it is unlcear why we test for them. The ocean_cycs code is based on [matlab code](http://www.soest.hawaii.edu/oceanography/faculty/zeebe_files/CO2_System_in_Seawater/csys.html) which only includes checks on temperature and salinity. It is unclear as to why we have those additional ranges in place the alkalinity check throws an error when in chem spinup = TRUE.
* chem spin up = TRUE 
     * throws an error even when using the old Hector default initial parameter values 
     * when the alk check is disabled we run into an error related to being unable to solve for roots. After some extensive testing, I have determined that that the problem is with the alkalinity. Based on the BGC paper alkalinity is defined after spinup is complete, when it is initialized it is set to 0, which causes the alkalinity assert that to be thrown and also the root finder to error. If we hard code the alkalinity (not something we want to do was using it as a test) we don't run into the error with the root finder. I suspect that this problem could be avoided if the alkalinity was not initialized with a value of 0 but I have not figured out where that is happening.

**Comment by kdorheim** [comment]
p.s. this work was made possible by emotional support from Scout 🐾

**Comment by bpbond** [comment]
Thanks for the great work @kdorheim , and appreciate the extra context @cahartin .

>With the new values in place is the carbon flow still around 100 Pg? 

This would be good to check, agreed. But I think getting the initial ocean C pools consistent with AR6 will be worth a change.

>chem spin up = TRUE

It seems like this is the major issue to deal with? Can we simply disallow spinup chemistry entirely? It's never been used as far as I know.

File: src/ocean_csys.cpp [hunk 1]
```diff
@@ -126,6 +126,7 @@ double find_largest_root( const int ncoeffs, double* a ) {
     PolyDerivFunctor polyFunctor(a, degree);
```
Comment (bpbond) [inline src/ocean_csys.cpp:126]: I bet there's a robust Boost function we could change to using and get rid of a bunch of code here. For v3.5?

Comment (bpbond) [reply src/ocean_csys.cpp:126]: https://www.boost.org/doc/libs/1_64_0/libs/math/doc/html/math_toolkit/roots.html

File: src/ocean_csys.cpp [hunk 2]
```diff
@@ -157,12 +162,15 @@ void oceancsys::ocean_csys_run( unitval tbox, unitval carbon )
     const double Tc = tbox.value( U_DEGC );
     const double Tk = Tc + 273.15;
 
-	// Check that all is OK with input data
-	H_ASSERT( Tk > 265 && Tk < 308, "bad Tk value" ); // Kelvin
-    H_ASSERT( dic > 1000e-6 && dic < 3700e-6, "bad dic value" );  // mol/kg
+    
+    // Check to make sure the system is within the range for K1 and K2.
```
Comment (bpbond) [inline src/ocean_csys.cpp:166]: 👍

File: src/ocean_component.cpp [hunk 3]
```diff
@@ -235,6 +218,46 @@ void OceanComponent::prepareToRun() {
...
+    const double LL_preind_C = LL_vol_frac * preind_C_surface;
+    const double HL_preind_C = HL_vol_frac * preind_C_surface;
+    const double I_preind_C = I_vol_frac * preind_C_ID;
+    const double D_preind_C = D_vol_frac * preind_C_ID;
+
+
+    // Set up our ocean box model.
+    H_LOG( logger, Logger::DEBUG ) << "Setting up ocean box model" << std::endl;
+   // surfaceHL.initbox( HL_preind_C, "HL" );
+       surfaceHL.initbox( unitval( 140, U_PGC ), "HL" );
```
Comment (bpbond) [inline src/ocean_component.cpp:244]: 😕 Why are we computing the preindustrial values above and then plugging in 140, etc. here?

**Review by bpbond — APPROVED** [review]
Good improvements here, @kdorheim —thank you. And holy cow I'm sorry I got so delayed on this.

**Comment by bpbond** [comment]
Let's follow up on CH's comment above, and check how big the high latitude -> deep box flow when exiting spinup.

**Comment by kdorheim** [comment]
Here is the output log from the first time step when spin up = 0 right after the active chemistry turns on 

```
Tue Feb 15 16:16:24 2022:DEBUG:log_state: ----- State of LL box -----
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    carbon = 844.184 -> 844.186
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    T=20.9 degC, surfacebox=1, active_chemistry=1
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    CarbonAdditions = 74.5659 (8.8329%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    CarbonSubtractions = 73.5659 (8.71444%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    FPgC = -0.997694 (-0.118184%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    ao_flux = 0 (0%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    oa_flux = 0.997694 (0.118184%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    K0 = 0.0324716 mol/l/atm Tr = 0.0337689 gC/m/month/atm
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    Surface DIC = 2236.67 umol/kg
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    Connection #0 to HL, k=0.0742532, window=1
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    Connection #1 to intermediate, k=0.0128912, window=1
Tue Feb 15 16:16:24 2022:DEBUG:log_state: ----- State of HL box -----
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    carbon = 151.208 -> 151.208
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    T=1.6 degC, surfacebox=1, active_chemistry=1
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    CarbonAdditions = 105.923 (70.0511%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    CarbonSubtractions = 106.923 (70.7124%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    FPgC = 1.00002 (0.661357%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    ao_flux = 1.00002 (0.661357%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    oa_flux = 0 (0%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    K0 = 0.0609349 mol/l/atm Tr = 0.0368937 gC/m/month/atm
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    Surface DIC = 2270.22 umol/kg
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    Connection #0 to deep, k=0.707124, window=1
Tue Feb 15 16:16:24 2022:DEBUG:log_state: ----- State of intermediate box -----
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    carbon = 9059.93 -> 9059.93
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    T=18 degC, surfacebox=0, active_chemistry=0
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    CarbonAdditions = 294.293 (3.24829%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    CarbonSubtractions = 294.293 (3.24829%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    K0 = 0 (undefined) Tr = 0 (undefined)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    Connection #0 to LL, k=0.0082303, window=1
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    Connection #1 to HL, k=0.0047726, window=1
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    Connection #2 to deep, k=0.01948, window=1
Tue Feb 15 16:16:24 2022:DEBUG:log_state: ----- State of deep box -----
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    carbon = 27969.5 -> 27969.5
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    T=18 degC, surfacebox=0, active_chemistry=0
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    CarbonAdditions = 283.41 (1.01328%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    CarbonSubtractions = 283.411 (1.01328%)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    K0 = 0 (undefined) Tr = 0 (undefined)
Tue Feb 15 16:16:24 2022:DEBUG:log_state:    Connection #0 to intermediate, k=0.0101328, window=1
Tue Feb 15 16:16:24 2022:DEBUG:record_state: Recording component state at t= 1746
```


It looks like `106.923` is moving from the HL box to the deep ocean box. So slightly larger than than the value of 100. @bpbond & @cahartin

File: src/ocean_component.cpp (hunk 3 above)
Comment (kdorheim) [reply src/ocean_component.cpp:244]: gah! good catch those are the old values being where as `// surfaceHL.initbox( HL_preind_C, "HL" );` is the code we want to commit but mistakenly pushed a version I was using to debug.

**Comment by kdorheim** [comment]
@bpbond and @cahartin in case you are interested in the values returned from the older values 

```
Tue Feb 15 16:30:46 2022:DEBUG:stashCValues: annualflux_sum=0.0032725 Pg C
Tue Feb 15 16:30:46 2022:DEBUG:log_state: ----- State of LL box -----
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    carbon = 783.646 -> 783.65
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    T=20.9 degC, surfacebox=1, active_chemistry=1
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    CarbonAdditions = 69.2904 (8.84205%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    CarbonSubtractions = 68.2904 (8.71444%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    FPgC = -0.9965 (-0.127162%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    ao_flux = 0 (0%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    oa_flux = 0.9965 (0.127162%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    K0 = 0.0324716 mol/l/atm Tr = 0.0337689 gC/m/month/atm
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    Surface DIC = 2076.28 umol/kg
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    Connection #0 to HL, k=0.0742532, window=1
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    Connection #1 to intermediate, k=0.0128912, window=1
Tue Feb 15 16:30:46 2022:DEBUG:log_state: ----- State of HL box -----
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    carbon = 140.525 -> 140.525
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    T=1.6 degC, surfacebox=1, active_chemistry=1
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    CarbonAdditions = 98.3685 (70.0008%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    CarbonSubtractions = 99.3685 (70.7124%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    FPgC = 0.999773 (0.711456%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    ao_flux = 0.999773 (0.711456%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    oa_flux = 0 (0%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    K0 = 0.0609349 mol/l/atm Tr = 0.0368937 gC/m/month/atm
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    Surface DIC = 2109.82 umol/kg
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    Connection #0 to deep, k=0.707124, window=1
Tue Feb 15 16:30:46 2022:DEBUG:log_state: ----- State of intermediate box -----
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    carbon = 8418.94 -> 8418.94
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    T=18 degC, surfacebox=0, active_chemistry=0
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    CarbonAdditions = 273.472 (3.2483%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    CarbonSubtractions = 273.472 (3.24829%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    K0 = 0 (undefined) Tr = 0 (undefined)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    Connection #0 to LL, k=0.0082303, window=1
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    Connection #1 to HL, k=0.0047726, window=1
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    Connection #2 to deep, k=0.01948, window=1
Tue Feb 15 16:30:46 2022:DEBUG:log_state: ----- State of deep box -----
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    carbon = 25991.8 -> 25991.8
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    T=18 degC, surfacebox=0, active_chemistry=0
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    CarbonAdditions = 263.369 (1.01328%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    CarbonSubtractions = 263.37 (1.01328%)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    K0 = 0 (undefined) Tr = 0 (undefined)
Tue Feb 15 16:30:46 2022:DEBUG:log_state:    Connection #0 to intermediate, k=0.0101328, window=1
```

The HL subtraction is 99, so much closer to the value of 100 that @cahartin mentioned above. However this is a 70.0% subtraction compared to a 70.7% subtraction could that be used to show that this was a trivial change?

**Comment by bpbond** [comment]
Seems like by either metric (99 vs. 100, 70.0% vs. 70.7%) it's a trivial change.

**Comment by kdorheim** [comment]
@bpbond [Here](https://rpubs.com/kdorheim/867919) is the markdown of how using the AR6 pre-industrial ocean C to initialize the Hector ocean carbon pools. 

Using the AR6 values the total ocean C increased by ~7% which causes a mean percent difference of 0.39% for pH, 14.91% for CO3, 7.66% for DIC, 7.6% in ocean C pools, 3.77% in sst, and less than 2% change in the big four. 

Comparing the with CMIP5 & observations the changes decrease the performance of pH. 

image

**Comment by bpbond** [comment]
Thanks @kdorheim - interesting. Interesting to see Hector move away from CMIP and HOTS! I don't think that's a reason to delay this PR though; better input data are (may be) exposing weaknesses in the model's algorithms.

**Comment by kdorheim** [comment]
@bpbond after passing github actions okay to merge?

**Pull request merged**